ProcessLib/CentralDifferencesJacobianAssembler.cpp

```diff
@@ -0,0 +1,170 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "CentralDifferencesJacobianAssembler.h"
+#include "BaseLib/Error.h"
+#include "MathLib/LinAlg/Eigen/EigenMapTools.h"
+#include "LocalAssemblerInterface.h"
+
+namespace ProcessLib
+{
+CentralDifferencesJacobianAssembler::CentralDifferencesJacobianAssembler(
+    std::vector<double>&& absolute_epsilons)
+    : _absolute_epsilons(std::move(absolute_epsilons))
+{
+    if (_absolute_epsilons.empty())
+        OGS_FATAL("No values for the absolute epsilons have been given.");
+}
+
+void CentralDifferencesJacobianAssembler::assembleWithJacobian(
+    LocalAssemblerInterface& local_assembler, const double t,
+    const std::vector<double>& local_x_data,
+    const std::vector<double>& local_xdot_data, const double dxdot_dx,
+    const double dx_dx, std::vector<double>& local_M_data,
+    std::vector<double>& local_K_data, std::vector<double>& local_b_data,
+    std::vector<double>& local_Jac_data)
+{
+    // TODO do not check in every call.
+    if (local_x_data.size() % _absolute_epsilons.size() != 0) {
+        OGS_FATAL(
+            "The number of specified epsilons (%u) and the number of local "
```

Comment:
These are not tolerances but perturbations applied to the local solution vector.